LL Platform JSON-RPC API documentation

pollastri-pierre opened the pull request:
This is first draft of API documentation for Ledger Live platform. 
The API is following JSON-RPC 2.0 standard.

It allows the following features:
- Request account
- Send transaction
- Receive transaction
- Open a bridge to a device to send APDUs
- Get access to crypto asset list value for the exchange app

In the documentation tags are used to describe the API sub-category (Accounts, Transaction, Device...) and the permission needed to use the method.

Review of the API can be done on OpenRPC playground for ease of read:
First draft: available [here](https://playground.open-rpc.org/?schemaUrl=https://raw.githubusercontent.com/pollastri-pierre/ledger-live-common/feature/platform_rpc/docs/platform-openrpc.json&uiSchema[appBar][ui:splitView]=false)

## Edit update 1
Update 1: available [here](https://playground.open-rpc.org/?schemaUrl=https://raw.githubusercontent.com/pollastri-pierre/ledger-live-common/9aef09dc7b16f418a37078ec4f1d7eed1d8eacfb/src/platform/openrpc.json&uiSchema[appBar][ui:splitView]=false)

Few changes have. been made on top of the first draft:
- All methods are now tagged with either `Version 0.0.0` or `Version 1.0.0`
- The file location has been changed to `src/platform/openrpc.json`
- New permision: PermissionRequestApp (allows request app and bridge app methods)
- New permission: PermissionRequestDashboard (allows to brige with the dashboard)
- Update description for minimumBalance in AccountSelection to specify that the amount is in crypto (smallest unit in the crypto like satoshi)
- Update description about freshAddress to explain what it means
- Change freshAddresses to freshAddress, third party application can now only see one account address instead of the full list
- Remove open_scheme method since it's not useful for v0 and v1

## Edit update 2
Update 2: available [here](https://playground.open-rpc.org/?schemaUrl=https://raw.githubusercontent.com/pollastri-pierre/ledger-live-common/0b3b9295944d94ed6b1d43b2bdf9f73055322d6d/src/platform/openrpc.json&uiSchema[appBar][ui:splitView]=false)

Changes:
- A `pluginName` in transaction object to be able to specify a plugin to use to verify&sign transaction on the device using a particular plugin

## Edit update 3
Update 3: available [here](https://playground.open-rpc.org/?schemaUrl=https://raw.githubusercontent.com/pollastri-pierre/ledger-live-common/11b1e2fe54be005916677ed82c420095b42a0e40/src/platform/openrpc.json&uiSchema[appBar][ui:splitView]=false)

Changes:
- Remove `pluginName`
- Add new object type DeviceAppFilter
- Add new parameter `forceApp` of type DeviceAppFilter to method `ledgerlive.account.sign_transaction`
- Add new object type SignedTransaction to store signed transaction data (for now just the raw transaction buffer)
- Change return type of method `ledgerlive.account.sign_transaction` to SignedTransaction

## Edit update 4
Update 4: available [here](https://playground.open-rpc.org/?schemaUrl=https://raw.githubusercontent.com/pollastri-pierre/ledger-live-common/f51cff3f4ec648659125c8e219175caf34b54d01/src/platform/openrpc.json&uiSchema[appBar][ui:splitView]=false)

Changes:
- Remove `request_app` from v0 and tag it in v1
- Remove `request_accounts` methods
- Add `get_accounts` methods to silently get all accounts
- Change `cryptoCurrencyId` (string) to `cryptoCurrencyIds` (list[string]) in AccountSelection object 
- Move `currencies.get` to v0
- Add currency name and currency ticker in currency object
- Remove send max from transaction object
- Remove subAccountId from transaction object
- Transaction object doesn't have `amount` and `recipient` in it, instead they are part of all Currency transaction objects
- Ethereum transaction: remove estimateFees and mode
- Bitcoin transaction: remove UTXO strategy
- Sign message method is now tagged as v0

## Edit update 5
Update 5: available [here](https://playground.open-rpc.org/?schemaUrl=https://raw.githubusercontent.com/pollastri-pierre/ledger-live-common/05d29ff49ed702e6d47fb3bd575a48fb20e2a88c/src/platform/openrpc.json&uiSchema[appBar][ui:splitView]=false)

Changes:
- Move broadcast transaction method to V0

pollastri-pierre committed:
Create platform-openrpc.json

gre commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+{
```
could it be possible to move this file under a new folder `src/platform/` ? and maybe `src/platform/openrpc.json`.
my understand is that we will have logic in live-common for it so i say let's land src/platform.
if we have doc i even like when it's collocated with the code, but my understanding is that this could even be used with code so i say we should collocate.

but what's your thought @IAmMorrow ?

@IAmMorrow  btw who can review this spec in our side? i say we land it asap and worse case is we make it evolve over time. we can don't need master to be stable if it's a new & unused src/platform.

pollastri-pierre replied:
Yep, I'll update.

lcastillo-ledger commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+    },
+    "methods": [
+        {
+            "name": "ledgerlive.open_scheme",
```
Do we have a list of schemes that can be opened from an iframe documented somewhere?

pollastri-pierre replied:
I don't think it's documented, but yep I would be interested on some documentation about this. 
@gre Do you  have some inputs ?

lcastillo-ledger commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+                }
+            ],
+            "result": {
+                "name": "address",
```
Would be amazing to have a signed by ledger version of this address in the long term

lcastillo-ledger commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+                {
+                    "name": "transaction",
+                    "required": false,
+                    "summary": "Default values for the transaction",
```
transaction field not required for a send? What happens if it's not specified?

lcastillo-ledger commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+        },
+        {
+            "name": "ledgerlive.account.broadcast",
+            "summary": "Use Ledger Live to broadcast a signed transaction",
```
broadcast vs send?

lcastillo-ledger commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+                },
+                {
+                    "name": "path",
+                    "required": true,
```
Why a path in addition to an account_id? Why not in sign_transaction?

pollastri-pierre replied:
Because in sign transaction, you are actually signing the transaction N times for N UTXO, each UTXO have it's own private key, you can't mix. Where as for sign message, you  sign it for  a given address not account (for BTC at least)

The path is not useful for single private key currencies

lcastillo-ledger commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+            "paramStructure": "by-name",
+            "tags": [
+                {
+                    "$ref": "#/components/tags/PermissionDeviceExchange"
```
Might be a specific permission for accessing the dashboard?

pollastri-pierre replied:
Indeed,

lcastillo-ledger commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+                        "$ref": "#/components/schemas/CryptoCurrencyIds"
+                    },
+                    "minimumBalance": {
+                        "$ref": "#/components/schemas/Amount"
```
In Fiat or in crypto?

lcastillo-ledger commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+                        "items": {
+                            "$ref": "#/components/schemas/Address"
+                        },
+                        "description": "An array containing all fresh addresses and paths"
```
What's a fresh address?

lcastillo-ledger commented on docs/platform-openrpc.json:
```diff
@@ -0,0 +1,1208 @@
+        },
+        {
+            "name": "ledgerlive.account.estimate_fees",
+            "summary": "Estimate fees for a given transaction. The method returns an updated version of the transaction passed in parameter",
```
So for a swap / send, current flow is the following:
1. Live calls exchange nano app to validate protobuf inputs and trigger a transaction with bitcoin nano app
2. Then Live communicates with BTC app to generate signed transaction and broadcast

I can see how to do number 1 with the current API (estimate_fees + curencies.get + exchange), but hiw would we do number 2?

gre commented:
I don't think there is a need to have the concept of "DeviceBridgeId" in the API. the id should be purely on LL side.
actually on desktop it's always `""` because it's a monolithic usb device.
on mobile, it's relevant but in fact it's only resolved on the UI at the last time so it's not something to transport in the api.

gre commented:
personally i would prefer a generic ledgerlive.account.prepareTransaction instead of "estimateFees" because prepareTransaction is the generic version of it

the idea to estimateFees feels too specific to a given coin (some coins don't even have fees) but actually the more general idea is to "resolve asynchronously something" which is what the bridge prepareTransaction idea was.

for ethereum for instance, prepareTransaction will:
- settle a gasPrice
- settle a gasLimit

also on top of this, prepareTransaction do some validation (maybe even in combination with "getTransactionStatus" that we could unify in one rpc) and it would be good to have such validation. because ultimately if prepareTransaction+getTransactionStatus pass => there is a very high chance all is good up to broadcast. that's the mindset on that bridge api.

https://github.com/LedgerHQ/ledger-live-common/blob/master/docs/gist-tx.md

lcastillo-ledger commented:
As long as you can easily access the transaction fields that would work as
a generic API.

In addition, if we go toward the device actions route, exposed through LL
API, then we do not need the fees themselves for Baanx: LL itself would do
the work of fees estimation when communicating with the exchange app.
- For now, they are needed because fees are transmitted directly to the
exchange app for validation on the nano screen.
- Since we intend to modify exchange app for supporting Baanx, I'm a bit
scared what it means in term of planning to add this in LL. Nice thing
about exposing APDUs is there is no modifications required in LL Core

On Tue, Apr 13, 2021 at 6:20 PM Gaëtan Renaudeau ***@***.***>
wrote:

> personally i would prefer a generic ledgerlive.account.prepareTransaction
> instead of "estimateFees" because prepareTransaction is the generic version
> of it
>
> the idea to estimateFees feels too specific to a given coin (some coins
> don't even have fees) but actually the more general idea is to "resolve
> asynchronously something" which is what the bridge prepareTransaction idea
> was.
>
> for ethereum for instance, prepareTransaction will:
>
>    - settle a gasPrice
>    - settle a gasLimit
>
> also on top of this, prepareTransaction do some validation (maybe even in
> combination with "getTransactionStatus" that we could unify in one rpc) and
> it would be good to have such validation. because ultimately if
> prepareTransaction+getTransactionStatus pass => there is a very high chance
> all is good up to broadcast. that's the mindset on that bridge api.
>
> https://github.com/LedgerHQ/ledger-live-common/blob/master/docs/gist-tx.md
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/LedgerHQ/ledger-live-common/pull/1134#issuecomment-818866653>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASXS5MAH2G7S7KXJHHSXRWTTIRVOBANCNFSM42VE3MNQ>
> .
>

-- 

Les informations contenues dans ce message électronique ainsi que celles 
contenues dans les documents attachés sont strictement confidentielles et 
sont destinées à l'usage exclusif du (des) destinataire(s) nommé(s).
Toute 
divulgation, distribution ou reproduction, même partielle, en est 
strictement interdite sauf autorisation écrite et expresse de l’émetteur.
Si vous recevez ce message par erreur, veuillez le notifier immédiatement à 
son émetteur par retour, et le détruire ainsi que tous les documents qui y 
sont attachés.


The information contained in this email and in any 
document enclosed is strictly confidential and is intended solely for the 
use of the individual or entity to which it is addressed.
Partial or total 
disclosure, distribution or reproduction of its contents is strictly 
prohibited unless expressly approved in writing by the sender.
If you have 
received this communication in error, please notify us immediately by 
responding to this email, and then delete the message and its attached 
files from your system.

IAmMorrow commented:
Nice job being done here !

A few comments:
Regarding `ledgerlive.open_scheme`, I'm not sure we should use an endpoint for that, just a ledgerlive:// link anywhere in the iframe would be good enough. Links can already be called from any external website without any restriction so i'm not sure it make a lot of sense to restrict them here.

I agree with the fact the the fees logic should be handled on ledger live side so we should not provide an estimate endpoint. However some dapps (like paraswap) will estimate the fees on their end and send you a transaction object with gasPrice and gasLimit already pre-filled. This doesn't prevent you from having the last word on it and to change it at your convenience but we need to keep that scenario in mind because currently, I fear it won't matter what fees the user is choosing on the paraswap UI since the live will most likely ignore them. Could be an issue

pollastri-pierre reviewed (commented):
> ## about estimateFees
> personally i would prefer a generic ledgerlive.account.prepareTransaction instead of "estimateFees" because prepareTransaction is the generic version of it
> 
> the idea to estimateFees feels too specific to a given coin (some coins don't even have fees) but actually the more general idea is to "resolve asynchronously something" which is what the bridge prepareTransaction idea was.
> 
> for ethereum for instance, prepareTransaction will:
> 
> * settle a gasPrice
> * settle a gasLimit
> 
> also on top of this, prepareTransaction do some validation (maybe even in combination with "getTransactionStatus" that we could unify in one rpc) and it would be good to have such validation. because ultimately if prepareTransaction+getTransactionStatus pass => there is a very high chance all is good up to broadcast. that's the mindset on that bridge api.
> 
> https://github.com/LedgerHQ/ledger-live-common/blob/master/docs/gist-tx.md

> 
> ## about permissions
> I tend to think we need to be more precise regarding things like list apps / request apps. "read device permission" is too large.
> I would vision something like:
> 
> * allow to list apps
> * allow to open apps [list of apps]
> * allowed to open any possible app
> 
> most of the daps will just be allowed to open app "Ethereum" and that would allow us to restrict more the fact only Ethereum app is authorized (e.g. when the device is no longer in that app, well it's a ❌ on the widget because you are no longer in the boundaries)
> 
> Same question regarding the accounts permissions. I feels it's important to either restrict to a given set of currencies.
> 

I agree, for the permission, I'll draft a proposition next week and we will be able to find tune  each type of permission  and specifically the parameters for each permission.
Basically we should have a request_account permission and  the manifest should state which currencies you are allowed to request


> ## about token accounts
> would we want to support it or not really? is Ethereum level only or is there a need to list the tokens?
> 

It will be useful for future versions for now I think we don't need it.

> ## about account data itself
> I'm not sure sure about the idea to list all possible freshAddresses. on the basic level it's just one.
> 

Done in  update 1 of the proposal

> ## about signTransaction / broadcast
> if we want to split into two parts (instead of just doing send) the current interface would not match: it should be a "signed operation" which is a bit more than a "signedTransaction" hex.
> This is why I tend to think we should just consider "send" and not bother splitting into two parts.

The issue is  that for  the first integration we really just need to sign transaction and for the second integration we will probably need  to create some metadata of the transaction let the third-party application do some work and then sign on Ledger Live the same transaction (like swap integration).
For the signed  operation do you have type you are thinking about (hex is the absolute minimum we need for first pass of integrations)

pollastri-pierre committed:
Add  an optional pluginName in Transaction object, to specify a plugin to use when signing transaction

pollastri-pierre committed:
Remove pluginName and add forceApp with DeviceAppFilter object and SignedTransaction object

pollastri-pierre committed:
Update 4:
- Remove `request_app` from v0 and tag it in v1
- Remove `request_accounts` methods
- Add `get_accounts` methods to silently get all accounts
- Change `cryptoCurrencyId` (string) to `cryptoCurrencyIds` (list[string]) in AccountSelection object
- Move `currencies.get` to v0
- Add currency name and currency ticker in currency object
- Remove send max from transaction object
- Remove subAccountId from transaction object
- Transaction object doesn't have `amount` and `recipient` in it, instead they are part of all Currency transaction objects
- Ethereum transaction: remove estimateFees and mode
- Bitcoin transaction: remove UTXO strategy
- Sign message method is now tagged as v0

pollastri-pierre committed:
Update 5  move broadcast transaction to V0 to be used during eth_sendTransaction RPC call

JunichiSugiura commented:
I accidentally approved it. How can i undo? 😅

gre commented:
unclear PR need and status. reopen if relevant.
